Feature/back end UI (#13)
plentymarkets/plenty-developers-docs: docs/en-gb/modules/back-end-ui/pages/overview.adoc

= Angular back end UI

This page gives an introduction to back end UIs and its components. All plentymarkets UIs are Angular UIs. Angular is a Javascript framework for creating web, desktop and mobile applications. Developing Angular applications involves the usage of Typescript along with HTML, CSS etc. Such applications are built by components, which define different areas in the UI that let you use specific UI functionalities. If you are not familiar with Angular yet and want to find out more, see the link:https://angular.io[Angular documentation].

== Back end menu information

All information about the back end menu entries is saved in a ui.json file. This file defines the entry points for the plugin. Views are then available at the respective menus.

This could look like the following:

.ui.json
[source,json]
----
include::example$ui.json[]
----

== Plugin menu entries

> The following section lists the specific properties of your plugin's menu entries.

Reviewer: Bei welchen Feldern handelt es sich um Pflichtangaben? Wenn alle Pflicht sind, würde ich das an dieser Stelle kurz erwähnen.

=== defaultEntryPoint

The default entry point of your plugin

=== namespace

The namespace of your plugin

=== menuEntries

The menu entries of your plugin. +

> The following information is saved for a menu entry:

Reviewer: Warum hast du hier eine Liste gewählt statt einer Tabelle?

> * `label`: The label to be displayed in the back end +

Reviewer: ```suggestion
* `label`: plentymarkets displays the label as menu entry in the back end.
```
Oder sowas in der Art, geht mir hauptsächlich darum, die Passiv-Konstruktion zu vermeiden.

> * `menu`: The menu that can be accessed by a plugin via entry points. +

Reviewer: ```suggestion
* `menu`: The menu the plugin accesses via an entry point.
```
s.o.

* `urlKey`: The route for the view in our plentymarkets back end, e.g.                         http://your-plentystore.co.uk/plenty/ui-backend/start/hello-world +

> *_Note:_* For the URL key of the plugin, only use hyphens. +

Reviewer: ```suggestion
*_Note:_* The URL key of the plugin may not contain any special characters except for hyphens.
```

* `entryPoint`: The entry point of your plugin +
* `icon`: The icon for the menu entry in the system tree. If no icon is specified, the standard plugin icon is displayed. To not display any icon, use `"icon": "none"`. Find a list of all icons <<ADJUST-LINK-terra-icons.adoc#,here>>.

== Integrating plugins into the back end

It is possible to seamlessly integrate your plugin into the plentymarkets back end. You can use entry points to add your plugin to the top navigation bar or routes to add it to the navigation tree of the *Setup* menu.

=== Menu entry points

> Back end views can be integrated in the plentymarkets main menu. Click <<reference-menu-entry-points.adoc#,here>> for a list of all menus that can be accessed by a plugin via entry points.

Reviewer: ```suggestion
Back end views can be integrated in the plentymarkets main menu. For a list of all menus that can be accessed by a plugin via entry points, refer to the xref:reference-menu-entry-points.adoc[Menu entry points] page.
```
Der Link sollte mehr als nur ein Wort umfassen.

Außerdem müssen Links zu anderen Seiten der Dokumentation mit xref formatiert werden.
https://docs.antora.org/antora/2.3/page/module-and-page-xrefs/

=== Routes in the system tree

> The Terra system tree offers plugin developers the opportunity to register their plugin behind a route. In doing so, the plugin becomes accessible in the system tree like a regular part of Terra. Click <<reference-routes-system-tree.adoc#,here>> for a list of all available routes that can be accessed by plugins. To find an instruction on how to register your plugin, see <<ADJUST-LINK-tutorials/backendview#code-uijson#,this page>>.

Reviewer: ```suggestion
The Terra system tree offers plugin developers the opportunity to register their plugin behind a route. In doing so, the plugin becomes accessible in the system tree like a regular part of Terra. For a list of all available routes that can be accessed by plugins, refer to the <<reference-routes-system-tree.adoc#, Routes in the system tree>> page. For instructions on how to register your plugin, refer to the xref:how-to-back-end-ui.adoc#code-uijson[back end UI guide].
```
s.o.

icon:warning[role="red"] [red]#Required# +
Note that all routes need an introducing `system`. Older routes beginning with `settings` will be mapped to match the new format.

> Since the new menu item for the system tree was renamed to 'Setup', the new full system tree path also starts with 'Setup'.

Reviewer: Hier wäre vielleicht eine Box besser?

=== Plugin positioning

> In addition to inserting plugins as desired in menus under *Setup*, it is also possible to determine their position depending on their route. If using a route not mentioned in <<reference-routes-system-tree.adoc#,this list>>, a parent node is created with the specified name. It appears underneath the actual entry. In case of several entries whose last part of the route is identical, they are all inserted underneath this part of the route. Thus, only the position of the first entry is considered.

Reviewer: ```suggestion
In addition to inserting plugins as desired in menus under *Setup*, it is also possible to determine their position depending on their route. If you use a route that's not mentioned in xref:reference-routes-system-tree.adoc[in the routes reference], a parent node is created with the specified name. It appears underneath the actual entry. In case of several entries whose last part of the route is identical, they are all inserted underneath this part of the route. Thus, only the position of the first entry is considered.
```
Kannst du den letzten Teil eventuell nochmal umformulieren? Ich bin mir nicht sicher, ob das so für alle verständlich ist.

The following example shows a basic `ui.json`:

.ui.json
[source,json]
----
include::example$ui.json[]
----

== Plugin menu entries

The following table lists the specific properties of your plugin's menu entries. +
icon:warning[role=red] [red]#Required# All entries listed below are mandatory.

[[table-plugin-menu-entries]]
.Menu entry properties
[cols="1,3"]
|===
|Property |Explanation

|*defaultEntryPoint*
|The default entry point of your plugin

|*namespace*
|The namespace of your plugin

|*menuEntries*
a|The menu entries of your plugin. +
 The following information is saved for a menu entry:

* `label`: plentymarkets displays the label as menu entry in the back end. +
* `menu`: The menu the plugin accesses via an entry point. +
* `urlKey`: The route for the view in our plentymarkets back end, e.g.                         http://your-plentystore.co.uk/plenty/ui-backend/start/hello-world. +
*_Note:_* The URL key of the plugin may not contain any special characters except for hyphens.
* `entryPoint`: The entry point of your plugin +

> * `icon`: The icon for the menu entry in the system tree. If no icon is specified, the standard plugin icon is displayed. To not display any icon, use `"icon": "none"`. Find a list of all icons on the xref:terra-icons.adoc[Terra icons] page.

Reviewer: Hatten wie nicht gesagt, dass wir die Icon-Seite auch erstmal weglassen?

Author: Die Icon-Seite soll bleiben. Nur das Tutorial "Using icons and components" kann raus.
